packages/backend/server/src/core/doc-service/service.ts

```diff
@@ -0,0 +1,67 @@
+import { Injectable, Logger } from '@nestjs/common';
+
+import {
+  Config,
+  CryptoHelper,
+  InternalServerError,
+  UserFriendlyError,
+} from '../../base';
+import { PgWorkspaceDocStorageAdapter } from '../doc';
+import { type DocRecord } from '../doc/storage';
+
+@Injectable()
+export class DocService {
+  private readonly logger = new Logger(DocService.name);
+
+  constructor(
+    private readonly config: Config,
+    private readonly workspace: PgWorkspaceDocStorageAdapter,
+    private readonly crypto: CryptoHelper
+  ) {}
+
+  async getDoc(workspaceId: string, docId: string): Promise<DocRecord | null> {
+    if (!this.config.docService.useRpc) {
+      return await this.workspace.getDoc(workspaceId, docId);
+    }
+    // fetch from controller
+    const url = `${this.config.docService.endpoint}/rpc/workspaces/${workspaceId}/docs/${docId}`;
+    const accessToken = `${docId},${this.crypto.sign(docId)}`;
+    try {
+      const response = await fetch(url, {
+        headers: {
+          'x-access-token': accessToken,
+        },
+      });
+      if (!response.ok) {
+        const text = await response.text();
+        if (response.status === 404) {
+          return null;
+        }
+        this.logger.error(
+          `Failed to fetch doc ${url}, status: ${response.status}, body: ${text}, headers: ${JSON.stringify(Object.fromEntries(response.headers.entries()))}`
+        );
+        throw new InternalServerError('Failed to fetch doc');
+      }
+      const timestamp = response.headers.get('x-doc-timestamp') as string;
+      const editor = response.headers.get('x-doc-editor-id') as string;
+      const bin = await response.arrayBuffer();
+      return {
+        spaceId: workspaceId,
+        docId,
+        bin: Buffer.from(bin),
```

Comment:
what the content would be if returns buffer in json body. it could respond stream and set the [timestamp] metadata in headers, and [editor] seems useless

Reply:
JSON itself does not support buffer data serialization, so custom deserialization must be done. Directly returning the raw buffer can reduce these serialization performance costs, which is particularly beneficial for extremely large doc buffers. That's why this solution was chosen; no matter how large the buffer data is, there is no problem.

Reply:
sry I was looking in the wrong code, this is the caller part.